inst/include/dust/r/dust.hpp

```diff
@@ -446,48 +446,61 @@ cpp11::sexp save_snapshots(const filter_state& snapshots, const T *obj,
   return(r_snapshots);
 }
 
-// TODO: this should be rewritten to return a list of 3 cpp1::sexp not
-// passing two as references.
 template <typename T>
-cpp11::writable::doubles run_filter(T * obj,
-                                    cpp11::sexp& r_trajectories,
-                                    cpp11::sexp& r_snapshots,
-                                    std::vector<size_t>& step_snapshot,
-                                    bool save_trajectories) {
+cpp11::sexp run_filter(T * obj, size_t step,
+                       std::vector<size_t>& step_snapshot,
+                       bool save_trajectories) {
   typename T::filter_state_type filter_state;
   cpp11::writable::doubles log_likelihood =
-    dust::filter::filter(obj, filter_state, save_trajectories, step_snapshot);
+    dust::filter::filter(obj, step, filter_state, save_trajectories,
+                         step_snapshot);
+  cpp11::sexp r_trajectories, r_snapshots;
   if (save_trajectories) {
     r_trajectories = dust::r::save_trajectories(filter_state.trajectories, obj);
   }
   if (!step_snapshot.empty()) {
-    r_snapshots = dust::r::save_snapshots(filter_state.snapshots, obj, step_snapshot);
+    r_snapshots = dust::r::save_snapshots(filter_state.snapshots, obj,
+                                          step_snapshot);
   }
-  return log_likelihood;
+
+  using namespace cpp11::literals;
+  return cpp11::writable::list({"log_likelihood"_nm = log_likelihood,
+                                "trajectories"_nm = r_trajectories,
+                                "snapshots"_nm = r_snapshots});
 }
 
 template <typename T, typename std::enable_if<!std::is_same<dust::no_data, typename T::data_type>::value, int>::type = 0>
-cpp11::sexp dust_filter(SEXP ptr, bool save_trajectories,
-                        cpp11::sexp r_step_snapshot) {
+cpp11::sexp dust_filter(SEXP ptr, SEXP r_step, bool save_trajectories,
+                        cpp11::sexp r_step_snapshot,
+                        cpp11::sexp min_log_likelihood) {
   T *obj = cpp11::as_cpp<cpp11::external_pointer<T>>(ptr).get();
   obj->check_errors();
 
   if (obj->n_data() == 0) {
     cpp11::stop("Data has not been set for this object");
   }
 
+  size_t step = std::numeric_limits<size_t>::max();
```

Comment:
Can't this just be left uninitialised as this value will never be read?

Reply:
Ah wait, I see how it's used. I wonder if `max_step` might be clearer here

Reply:
updated that now, definitely clearer